cluster-up/cluster/kind-k8s-sriov-1.17.0/certcreator/certsecret.go

```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"kubevirt.io/kubevirtci/cluster-up/cluster/kind-k8s-sriov-1.17.0/certcreator/certlib"
+)
+
+func handleKubeClientConfig(kubeconfig string) (*rest.Config, error) {
+	if kubeconfig == "" {
+		log.Printf("Using env kubeconfig %s", kubeconfig)
+		kubeconfig = os.Getenv("KUBECONFIG")
+	}
+
+	var config *rest.Config
+	var err error
+	if kubeconfig != "" {
+		log.Printf("Loading kube client config from path %q", kubeconfig)
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	} else {
+		log.Printf("Using in-cluster kube client config")
+		config, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not get the client: %v", err)
+	}
+
+	return config, nil
+}
+
+func generate(hookName, namespace string) ([]byte, []byte, error) {
+	serviceName := strings.Join([]string{hookName, "service"}, "-")
+
+	certConfig := certlib.SelfSignedCertificate{
+		CommonName: strings.Join([]string{serviceName, namespace, "svc"}, "."),
+		DNSNames: []string{
+			serviceName,
+			strings.Join([]string{serviceName, namespace}, "."),
+			strings.Join([]string{serviceName, namespace, "svc"}, ".")},
+	}
+	err := certConfig.Generate()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate self-signed certificate: %v", err)
+	}
+	log.Printf("Self-Signed certificate created sucessfully for CN %s", certConfig.CommonName)
+
+	return certConfig.Certificate.Bytes(), certConfig.PrivateKey.Bytes(), nil
+}
+
+func exportCertificateFile(data []byte, filePath string) error {
+	certificateFileName := fmt.Sprintf("%s.cert", filePath)
+	encodedData := []byte(base64.StdEncoding.EncodeToString(data))
+	if err := ioutil.WriteFile(certificateFileName, encodedData, 0644); err != nil {
+		return fmt.Errorf("failed to write content to file %s: %v", filePath, err)
+	}
+	log.Printf("certificate exported successfully to: %s", filePath)
+
+	return nil
+}
+
+func createSecret(clusterApi kubernetes.Interface, namespace, secretName string, certificate, key []byte) error {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: namespace,
+		},
+		Data: map[string][]byte{
+			"tls.crt": certificate,
+			"tls.key": key,
+		},
+	}
+
+	err := wait.Poll(time.Second*5, time.Minute*3, func() (bool, error) {
+		_, createSecretError := clusterApi.CoreV1().Secrets(namespace).Create(secret)
+		if createSecretError != nil {
+			return false, createSecretError
```

Comment:
if you return err it will stop poll, isnt it ? (see please function description)
i am not sure this what you wanted right?

Reply:
Agree, also why do we need the Poll?

Reply:
There were etcd timeouts

Reply:
During sriov-operator deployment right after all pods are created we get ectd timeout errors often,
we got those erorrs during CSR or at secret creation.
We still investigating it and try to find the root cause for that.
Using a Poll here is a woraround to make the provider more stable and reduce job failures on CI.

Reply:
https://github.com/kubevirt/kubevirt/issues/4519#issuecomment-724913645
Thats what i suggest to try

I would split it to two orhtogonal parts at least for now
1. Retry here looks fine until [2] would fully resolved before its self contained
2. Fix the issue https://github.com/kubevirt/kubevirt/issues/4519, and then [1] can be reconsidered, but not before that imo.

Reply:
I dont understand what is [1], [2] and "But to fix according first comment"
This change is what discussed offline and on this PR https://github.com/kubevirt/kubevirtci/pull/468

Reply:
Lets focus please on the bug here
from what i know (and Quique agreed above) you shouldnt return error in `wait.Poll` 
in case you want it to retry, which is what you want here.
error is one of the breaking condition of it.
see function documentation please.

Reply:
Done